apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/InitServiceNameInstrumentation2.java

```diff
@@ -0,0 +1,112 @@
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+/**
+ * Instruments
+ * <ul>
+ *     <li>{@link javax.servlet.ServletContainerInitializer#onStartup(java.util.Set, javax.servlet.ServletContext)} </li>
+ *     <li>{@link jakarta.servlet.ServletContainerInitializer#onStartup(java.util.Set, jakarta.servlet.ServletContext)}</li>
+ * </ul>
+ * <p>
+ * Determines the service name based on the webapp's {@code META-INF/MANIFEST.MF} file early in the startup process.
+ * As this doesn't work with runtime attachment, the service name is also determined when the first request comes in.
+ */
+public abstract class InitServiceNameInstrumentation2 extends AbstractServletInstrumentation {
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        return nameContains("Initializer");
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return not(isInterface()).and(hasSuperType(namedOneOf(
+            "javax.servlet.ServletContainerInitializer", "jakarta.servlet.ServletContainerInitializer")));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("onStartup")
+            .and(takesArguments(2))
+            .and(takesArgument(0, Set.class))
+            .and(takesArgument(1, nameEndsWith("ServletContext")));
+    }
+
+    public static class JavaxInitServiceNameInstrumentation extends InitServiceNameInstrumentation2 {
+
+        private static final JavaxServletApiAdapter adapter = JavaxServletApiAdapter.get();
+
+        @Override
+        public String rootClassNameThatClassloaderCanLoad() {
+            return "javax.servlet.AsyncContext";
+        }
+
+        public static class AdviceClass {
+
+            @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+            public static void onEnter(@Advice.Argument(1) @Nullable Object servletContext) {
+                if (servletContext instanceof javax.servlet.ServletContext) {
+                    ServletServiceNameHelper.determineServiceName(adapter, (javax.servlet.ServletContext) servletContext, tracer);
+                }
+            }
+        }
+    }
+
+    public static class JakartaInitServiceNameInstrumentation extends InitServiceNameInstrumentation2 {
+
+        private static final JakartaServletApiAdapter adapter = JakartaServletApiAdapter.get();
+
+        @Override
+        public String rootClassNameThatClassloaderCanLoad() {
+            return "jakarta.servlet.AsyncContext";
+        }
+
+        public static class AdviceClass {
+
+            @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+            public static void onEnter(@Advice.Argument(1) @Nullable Object servletContext) {
+                if (servletContext instanceof javax.servlet.ServletContext) {
```

Comment:
copy/paste error this should be `jarkata.servlet.ServletContext` and you can also make the parameter of this type instead of `Object.

Reply:
Doh, I will fix it.